### PR TITLE
Introduce major version pinning for mlflow requirement in saved models

### PR DESCRIPTION
**Pull request by BenWilson2**
Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Add major version pinning support for the mlflow package requirement within saved models

## How is this patch tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Introduce major version pinning for mlflow requirement in saved models by specifying major version ranges based on the version of the client installation of MLflow used when a model is saved

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/environment.py [hunk 1]
```diff
@@ -442,11 +445,40 @@ def _is_mlflow_requirement(requirement_string):
...
+            # from source.
+            # Example: mlflow @ git+https://github.com/mlflow/mlflow@branch_2.0
+            repository_matches = ["/mlflow", "mlflow@git"]
+
+            return any(
+                match in requirement_string.replace(" ", "").lower() for match in repository_matches
+            )
+
+
+def _generate_mlflow_version_pinning(strict=False):
```
Comment (harupy) [inline mlflow/utils/environment.py:459]: Do we need `strict`? Do we set `strict` to `True` somewhere?

Comment (BenWilson2) [reply mlflow/utils/environment.py:459]: No, I left it in here as a question for the PR review - should we do:
a) major pinning
b) minor pinning
c) strict micro pinning
Thoughts?
If a) -> I'll remove the arg and simplify the function.

Comment (dbczumar) [reply mlflow/utils/environment.py:459]: Awesome, just major pinning. We should use `mlflow<3` for all models logged in MLflow 2.0. Note that `mlflow~=2` is actually error prone due to quirks in Python / PyPI's dependency resolution procedure, so `mlflow<3` is preferred.

Comment (BenWilson2) [reply mlflow/utils/environment.py:459]: awesome! I'll make that change :)

File: mlflow/utils/environment.py [hunk 2]
```diff
@@ -442,11 +444,31 @@ def _is_mlflow_requirement(requirement_string):
...
+def _generate_mlflow_version_pinning():
+    """
+    Determines the current MLflow version that is installed and adds a pinned boundary version range
+    for mlflow. The upper bound is a cap on the next major revision. The lower bound is a cap on
+    the current installed minor version(i.e., 'mlflow<3,>=2.1')
+    :return: string for MLflow dependency version
+    """
+    mlflow_environment_version = Version(VERSION)
+    current_major_version = mlflow_environment_version.major
+    current_minor_version = mlflow_environment_version.minor
```
Comment (harupy) [inline mlflow/utils/environment.py:467]:
```suggestion
    mlflow_version = Version(VERSION)
    current_major_version = mlflow_version.major
    current_minor_version = mlflow_version.minor
```

Nit: can we remove `_environment`? `mlflow_environment_version` sounds like the version of an environment where mlflow is installed.

**Review by harupy — APPROVED** [review]
LGTM!

File: mlflow/utils/environment.py (hunk 2 above)
Comment (BenWilson2) [reply mlflow/utils/environment.py:467]: of course! Pushing the change in a few minutes :)

**Pull request merged**